soroban-env-host/src/host/metered_vector.rs

```diff
@@ -1,250 +1,298 @@
+            if err.is_some() {
+                return Ordering::Equal;
+            }
+            match cmp(probe) {
+                Ok(ord) => ord,
+                Err(he) => {
+                    err = Some(he);
+                    Ordering::Equal
+                }
+            }
+        });
+        match err {
+            Some(he) => Err(he),
+            None => Ok(res),
+        }
     }
 
-    /// Time: O(1)
-    #[inline]
-    pub fn iter(&self) -> Iter<'_, A> {
+    pub fn iter(&self) -> std::slice::Iter<'_, A> {
         self.vec.iter()
     }
 
-    /// Time: O(1)
-    #[inline]
-    pub fn iter_mut(&mut self) -> IterMut<'_, A> {
+    pub fn iter_mut(&mut self) -> std::slice::IterMut<'_, A> {
         self.vec.iter_mut()
     }
 }
 
-impl<A: Clone> Clone for MeteredVector<A> {
-    fn clone(&self) -> Self {
-        Self {
-            budget: self.budget.clone(),
-            vec: self.vec.clone(),
-        }
-    }
-}
-
-impl<A: Clone> MeteredClone for MeteredVector<A> {
-    fn metered_clone(&self, budget: &Budget) -> Result<Self, HostError> {
-        assert!(Rc::ptr_eq(&self.budget.0, &budget.0));
-        self.charge_new()?;
-        Ok(self.clone())
-    }
-}
-
-impl<A: Clone + PartialEq> PartialEq for MeteredVector<A> {
-    fn eq(&self, other: &Self) -> bool {
-        Rc::ptr_eq(&self.budget.0, &other.budget.0) && self.vec == other.vec
-    }
-}
-
-impl<A: Clone + Eq> Eq for MeteredVector<A> {}
-
-impl<A: Clone + PartialOrd> PartialOrd for MeteredVector<A> {
-    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-        assert!(Rc::ptr_eq(&self.budget.0, &other.budget.0));
-        self.vec.partial_cmp(&other.vec)
+impl<A> MeteredClone for MeteredVector<A>
+where
+    A: MeteredClone,
+{
+    fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
+        self.charge_new(self.len(), budget)
     }
 }
 
-impl<A: Clone + Ord> Ord for MeteredVector<A> {
-    fn cmp(&self, other: &Self) -> Ordering {
-        assert!(Rc::ptr_eq(&self.budget.0, &other.budget.0));
-        self.vec.cmp(&other.vec)
+impl<Elt: MeteredClone> Compare<MeteredVector<Elt>> for Budget
+where
+    Budget: Compare<Elt, Error = HostError>,
+{
+    type Error = HostError;
+
+    fn compare(
+        &self,
+        a: &MeteredVector<Elt>,
+        b: &MeteredVector<Elt>,
+    ) -> Result<Ordering, Self::Error> {
+        self.as_budget()
+            .charge(CostType::VecEntry, a.vec.len().min(b.vec.len()) as u64)?;
+        <Self as Compare<Vec<Elt>>>::compare(self, &a.vec, &b.vec)
     }
 }
 
-impl<A: Clone + Ord> MeteredCmp for MeteredVector<A> {
-    fn metered_cmp(&self, other: &Self, budget: &Budget) -> Result<Ordering, HostError> {
-        assert!(Rc::ptr_eq(&self.budget.0, &other.budget.0));
-        self.charge_cmp(min(self.len(), other.len()) as u64)?;
-        Ok(self.vec.cmp(&other.vec))
+impl<Elt: MeteredClone> Compare<MeteredVector<Elt>> for Host
+where
+    Host: Compare<Elt, Error = HostError>,
```

Comment:
I think these two impls can be replaced by a generic:

`impl<B: AsBudget, Elt: MeteredClone> Compare<MeteredVector<Elt>> for B
where
    B: Compare<Elt, Error = HostError>,`

Reply:
I tried this but I get one of the perplexing trait errors: 

```
error[E0210]: type parameter `B` must be covered by another type when it appears before the first local type (`MeteredVector<Elt>`)
   --> soroban-env-host/src/host/metered_vector.rs:266:6
    |
266 | impl<B: AsBudget, Elt: MeteredClone> Compare<MeteredVector<Elt>> for B
    |      ^ type parameter `B` must be covered by another type when it appears before the first local type (`MeteredVector<Elt>`)
    |
    = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local, and no uncovered type parameters appear before that first local type
    = note: in this case, 'before' refers to the following order: `impl<..> ForeignTrait<T1, ..., Tn> for T0`, where `T0` is the first and `Tn` is the last
```

I've found basically that half or 2/3 of the times when I try to do something "more generic" the trait system prevents me from doing it, either due to one of these sorts of prohibitions or else by triggering infinite recursion or some other solver-conflict. @leighmcculloch has had a similar experience, and we've been leaning towards "more impls that are repetitive but at least work" in a lot of recent code.

Reply:
Thanks for the explaination.